File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    let result = 0;
    let arr = s.split('');
    
    const obj = {
        ')': '(', 
        '}': '{', 
        ']': '['
    };
    
    for (let i = 0; i < arr.length; i++) {
        if (i) arr.push(arr.shift());
        
        const stack = [];
        const temp = [];
        for (let str of arr) {
            if (str === '(' || str === '{' || str === '[') stack.push(str);
            else {
                if (stack.length && stack[stack.length-1] === obj[str]) stack.pop();
                else temp.push(str);
            }
        }
        
        if (!stack.length && !temp.length) result++;
    }
    return result;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    result = 0
    dict_s = { ')': '(', ']': '[', '}': '{' }
    opened_s = ['(', '[', '{']
    closed_s = [')', ']', '}']
    
    for i in range(len(s)):
        new_s = s[i:] + s[:i]
        
        if new_s[0] in closed_s: continue
        
        stack = []
        for n in new_s:
            if len(stack) == 0 or n in opened_s:
                stack.append(n)
            elif dict_s[n] == stack[-1]:
                stack.pop()
        
        if len(stack) == 0:
            result += 1
    
    return result